ViewEnvironmentUI/Sources/ViewEnvironmentPropagatingObject.swift

```diff
@@ -0,0 +1,272 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import ViewEnvironment
+
+/// A protocol describing a ``ViewEnvironmentPropagating`` object that can:
+/// - Override the ancestor and descendants
+/// - Add environment update observations
+/// - Flag the backing object for needing to have the `ViewEnvironment` reapplied (e.g. `setNeedsLayout()`)
+///
+/// This protocol was abstracted to share propagation logic between `UIViewController` and `UIView`'s support for
+/// ``ViewEnvironmentPropagating``, but could be used for any object-based node that wants to support
+/// `ViewEnvironment` propagation.
+///
+public protocol ViewEnvironmentPropagatingObject: AnyObject, ViewEnvironmentPropagating {
+    /// The default ancestor for `ViewEnvironment` propagation.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var defaultEnvironmentAncestor: ViewEnvironmentPropagating? { get }
+
+    /// The default descendants for `ViewEnvironment` propagation.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var defaultEnvironmentDescendants: [ViewEnvironmentPropagating] { get }
+
+    /// Informs the backing object that this specific node should be flagged for another application of the
+    /// `ViewEnvironment`.
+    ///
+    /// For `UIViewController`/`UIView`s this typically corresponds to `setNeedsLayout()`.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    func setNeedsApplyEnvironment()
+}
+
+extension ViewEnvironmentPropagating where Self: ViewEnvironmentPropagatingObject {
+    /// Consumers _must_ call this function when the environment should be re-applied, e.g. in
+    /// `viewWillLayoutSubviews()` for `UIViewController`s and `layoutSubviews()` for `UIView`s.
+    ///
+    /// This will call ``apply(environment:)`` on the receiver if the node has been flagged for needing update.
+    ///
+    public func applyEnvironmentIfNeeded() {
+        guard needsEnvironmentUpdate else { return }
+
+        needsEnvironmentUpdate = false
+
+        if let observing = self as? ViewEnvironmentObserving {
+            let environment = observing.environment
+            observing.apply(environment: environment)
+        }
+    }
+}
+
+extension ViewEnvironmentPropagatingObject {
+    /// Adds a `ViewEnvironment` change observation.
+    ///
+    /// The observation will only be active for as long as the returned lifetime is retained or
+    /// `cancel()` is called on it.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public func addEnvironmentNeedsUpdateObserver(
```

Comment:
Do we still have a use case for this?

Reply:
I'm still using it in the prototype inside of `TrampolineModalListObserver`.

I think we may actually want to be using this method of observation instead of overriding `environmentDescendantsOverride` when the logic making this observation is not owned by the view controller itself.

For example, I'm currently playing with having `WorkflowHostingController` set its `environmentDescendantsOverride` to an empty array to prevent vanilla propagation invalidations causing a re-render of it's owned `rendering` since we'll already be performing a re-render on `environmentDidChange`. In our `TrampolineModalPresenter`, we need some way to observe invalidations triggered by its `owner`, but that owner could be a `WorkflowHostingController`. Rather than trying to override `environmentDescendantOverrides` and having to compose the existing override (similar to calling `super`), it might be cleaner to consider `environmentDescendantsOverride` something that only the VC it's applied to should override, and if you need to hook into invalidations, you use the `addEnvironmentNeedsUpdateObserver(...)`.

The composing approach does seem to work, but seems like it could get hard to follow when debugging and potentially difficult to document, but I'd be open to utilizing that approach for cases like this if we'd rather consolidate the API a bit.